src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs

```diff
@@ -20,6 +20,31 @@ public static bool CanCastTo(this TypeDesc thisType, TypeDesc otherType)
             return thisType.CanCastToInternal(otherType, null);
         }
 
+        public static bool IsCompatibleWith(this TypeDesc thisType, TypeDesc otherType)
+        {
+            // Structs can be cast to the interfaces they implement, but they are not compatible according to ECMA I.8.7.1
+            bool isCastFromValueTypeToReferenceType = otherType.IsValueType && !thisType.IsValueType;
+            if (isCastFromValueTypeToReferenceType)
+            {
+                return false;
+            }
+
+            // Nullable<T> can be cast to T, but this is not compatible according to ECMA I.8.7.1
+            bool isCastFromNullableOfTtoT = thisType.IsNullable && otherType.IsEquivalentTo(thisType.Instantiation[0]);
+            if (isCastFromNullableOfTtoT)
+            {
+                return false;
+            }
+
+            return otherType.CanCastTo(thisType);
```

Comment:
This is still going to throw if thisType or otherType is a pointer or function pointer. Do we need to handle that? (I had that feedback in https://github.com/dotnet/runtime/pull/35308#discussion_r420007424 but it probably got lost.

Reply:
Ah, let me add those to the test suite. It seems pointers cannot be compatible-with each other unless they are the same. I am actually not sure if the function pointers correspond to "method-signature compatible-with" case in the ECMA spec.